### PR TITLE
[util] New i2csvg utility to make pictures of I2C transactions

### PR DESCRIPTION
**Pull request by mdhayter**
Add a new util to convert text representations of the data
written to the master I2C fifo into pictures matching those in
the SMBus specification.

i2csvg can work from either a readable text format or from a
hex dump of values written to the fifo. See the README.md for
details and examples.

Used to show how all the SMBus 3.0 chapter 6.7 commands can
be encoded using the interface. Added markdown that can be
merged into the I2C documentation eventually.

Signed-off-by: Mark Hayter <mark.hayter@gmail.com>

### CODE REVIEW

**Comment by mdhayter** [comment]
I wanted to check the I2C interface would work with SMBus commands and ended up writing a little utility to make svgs from what would be written to the I2C master fifo.

In addition to regular review, I have some questions:

1) Not sure if it belongs in util (where I have put it) or should be somewhere under ip/i2c?
@asb @sjgitty suggestions?

2) Once PR #1047 is merged (I2C docs) the .md and svgs generated from the smbus subdirectory should probably be in the documentation in ip/i2c/doc
@martin-lueker Im happy to do this or for you to do it

3) In the olden days I would have made a patch to docgen to allow inline I2C descriptions. At this point I don't understand hugo enough to do this. If people think it might be useful I can look in to it.

**Comment by sjgitty** [comment]
This is nice, thanks for adding @mdhayter 

> 1. Not sure if it belongs in util (where I have put it) or should be somewhere under ip/i2c?
>    @asb @sjgitty suggestions?

`util/` makes sense.

> 2. Once PR #1047 is merged (I2C docs) the .md and svgs generated from the smbus subdirectory should probably be in the documentation in ip/i2c/doc
>    @martin-lueker Im happy to do this or for you to do it

would be great to have @martin-lueker @igk8  and @tunghoang290780 review this

> 3. In the olden days I would have made a patch to docgen to allow inline I2C descriptions. At this point I don't understand hugo enough to do this. If people think it might be useful I can look in to it.

that could be added as a second pass I imagine.

File: util/i2csvg/README.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+For simple examples text can be piped in to the tool using echo.
+
+A simple I2C write consists of a Start with Address and direction bit of zero, and a single data byte (4 in the example) with a Stop.
+
+The baseline tool will generate an svg file (which can be viewed in a browser):
+
+```console
+$ cd $REPO_TOP/util
+$ echo "SA0, P4" | ./i2csvg.py > out.svg
+```
```
Comment (sjgitty) [inline util/i2csvg/README.md]: I assume you want the text in addition to the "table row"? I.e. in this case the text is

`S.... A0, .P... 0x4`

File: util/i2csvg/README.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+If svg generation is not suppressed then the text or debug output and the svg are generated in an HTML file.
+
+```console
+$ cd $REPO_TOP/util
+$ echo "SA0, P4" | ./i2csvg.py --text > out.html
+```
+
+The smbus directory contains a files with each of the commands listed in SMBus 3.0 specification.
```
Comment (sjgitty) [inline util/i2csvg/README.md]: "contains files"

File: util/i2csvg/README.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
+# i2csvg -- Generate svg pictures of i2c commands from text description
+
+i2csvg is a Python tool to read i2c transactions from a text file and
+generate `svg` pictures of them.
+
+Two source representations are supported.
+The first is a more readable format and describes transactions in text.
+The second is a hex-dump of values written to the I2C fifo.
+In either format a line starting with a `#` is taken as a comment and a line staring with a `T` will be used as a title in the output.
```
Comment (sjgitty) [inline util/i2csvg/README.md]: We're attempting to standardize on 80 char READMEs, one sentence per line on other `.md` files.

File: util/i2csvg/README.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+To assist in parsing debug output the tool can be given a prefix.
+Only lines with the prefix will be processed (after removal of the prefix).
+
+### Textual Source Format
+
+The textual description includes the flags and value written to the I2C host interface fifo to make the transaction.
```
Comment (sjgitty) [inline util/i2csvg/README.md]: I believe @imphil made an attempt to rally around `FIFO` as opposed to `fifo`

File: util/i2csvg/README.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+The smbus directory contains a files with each of the commands listed in SMBus 3.0 specification.
+These can be used to generate an HTML file with all the transactions expanded.
+
+```console
+$ cd $REPO_TOP/util
+$ ./i2csvg.py i2csvg/smbus/*.txt > out.html
+```
+
+Alternatively, these can all be converted to individual svg files (for example for inclusion in documents) with the --multiout command.
+The output filename matches the input filename with the extension replaced with .svg
```
Comment (sjgitty) [inline util/i2csvg/README.md]: Need trailing `.` on these two lines.

File: util/i2csvg/README.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+Note that if the --text or --debug flags are given then the output and extension will be html
```
Comment (sjgitty) [inline util/i2csvg/README.md]: probably want all command line arguments in `backquote` like `--text` and `--debug`

File: util/i2csvg/README.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+03-ReceiveByte.txt  05-ReadWord.txt     08-BlockWrRdPCall.txt  12-Write64.txt
+04-WriteByte.svg    06-ProcessCall.svg  09-HostNotify.svg      13-Read64.svg
+04-WriteByte.txt    06-ProcessCall.txt  09-HostNotify.txt      13-Read64.txt
+```
+
+### Example use in Debugging
+
+The examples directory contains example output from a microcontroller using the I2C interface.
+Mixed in to the other debug output are lines produced when writes are done to the I2C fifo.
+These lines are prefixed by I2CF:
```
Comment (sjgitty) [inline util/i2csvg/README.md]: Maybe: "These lines are prefixed by `I2CF:`." ?

File: util/i2csvg/README.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
+# i2csvg -- Generate svg pictures of i2c commands from text description
+
+i2csvg is a Python tool to read i2c transactions from a text file and
```
Comment (sjgitty) [inline util/i2csvg/README.md]: usually you have `I2C` but here and above you have `i2c`, probably want to standardize on `I2C`?

File: util/i2csvg/README.md (hunk 7 above)
Comment (martin-lueker) [reply util/i2csvg/README.md]: Or better yet:
`These lines are prefixed by "I2CF:". `

File: util/i2csvg/README.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+|C | With R, allows read to continue by ACK of the last byte read |
+|N | NACK OK, do not raise an error on a NACK |
+
+The value may be an integer (binary, octal, decimal and hex supported) or one of the special values that change how the output is drawn:
+
+|Value | Use |
+|------|-----|
+|A0 | Address with direction bit set to 0 (write) |
+|A1 | Address with direction bit set to 1 (read) |
+|A2 | Address in Quick command with direction bit used for data |
```
Comment (martin-lueker) [inline util/i2csvg/README.md]: @mdhayter: I'm unfamiliar with the "Quick command" format.  Is this an SMBus term?  As far as I can tell it's not in the I2C base spec.

File: util/i2csvg/README.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+$ cd $REPO_TOP/util
+$ echo "SA0, P4" | ./i2csvg.py > out.svg
+```
+
+The text version (which is just the input with more structured formatting) can be viewed in the terminal by suppressing the svg output:
+
+```console
+$ cd $REPO_TOP/util
+$ echo "SA0, P4" | ./i2csvg.py --text --nosvg
+S.... A0
```
Comment (martin-lueker) [inline util/i2csvg/README.md]: What exactly do the dots mean?  I see a lot of transactions with dots or with out dots.  Is the idea to make an optional 5-bit string to match the 5-flags?

File: util/i2csvg/examples/SMBusCommands.txt [hunk 11]
```diff
@@ -0,0 +1,102 @@
+T 6.5.1 Quick Command
+SPA2
+
+T 6.5.2 Send Byte
+SA0
+P'Data'
+
+T 6.5.3 Receive Byte
+SA1
+RP1
```
Comment (martin-lueker) [inline util/i2csvg/examples/SMBusCommands.txt]: Question: Do the flags have to be in any particular order?

File: util/i2csvg/README.md (hunk 9 above)
Comment (mdhayter) [reply util/i2csvg/README.md]: Yes, this is an SMBus thing. 6.5.1 in http://smbus.org/specs/SMBus_3_0_20141220.pdf

I worked through all the commands in that chapter and all seem possible (the only quirk being a split in the ones that need to read N part way though)

File: util/i2csvg/README.md (hunk 10 above)
Comment (mdhayter) [reply util/i2csvg/README.md]: On output I pretty-print line up the flags and put dots where they are not set, which of course also aligns the value (assuming fixed-width font). Could change to spaces but dots looked better to me.

File: util/i2csvg/examples/SMBusCommands.txt (hunk 11 above)
Comment (mdhayter) [reply util/i2csvg/examples/SMBusCommands.txt]: On input flags can be in any order (and dots are ignored so the pretty-print output could be used as input).

Should add a comment in the text.

File: util/i2csvg/README.md (hunk 7 above)
Comment (mdhayter) [reply util/i2csvg/README.md]: used backquotes

File: util/i2csvg/README.md (hunk 3 above)
Comment (mdhayter) [reply util/i2csvg/README.md]: wrapped

File: util/i2csvg/examples/SMBusCommands.txt (hunk 11 above)
Comment (mdhayter) [reply util/i2csvg/examples/SMBusCommands.txt]: comment added

**Comment by mdhayter** [comment]
Fixed all comments. The requested rewrap has made it harder to see them but I checked them off in the comments as I did them.

**Comment by martin-lueker** [comment]
LGTM

**Comment by mdhayter** [comment]
I don't have write access, can someone merge this?

**Pull request merged**